### ISSUE TITLE
Include support for dynamic arrays in abstract code

### ISSUE THREAD

**Issue by thesamovar** [brian-team/brian2#431]
If we did this, could it be the case that there might be big simplifications to many templates (and maybe even a number of templates could be removed)?

I'm putting this up as a question because there might be hidden problems that I haven't thought about (which might indeed be why we didn't do it originally!).

**Comment by mstimberg** [comment]
What exactly do you have in mind here? Reading/writing of dynamic array entries is of course already possible in abstract code, so you are talking about resizing them? Or is this about 2-dimensional arrays (such as in `StateMonitor`)?

**Comment by thesamovar** [comment]
Yeah I mean resizing, the idea being to simplify monitors, creation of synapses, etc.

**Comment by mstimberg** [comment]
I still don't quite see how this could fit in with abstract code syntax. Abstract code is expressed as something that is independent of data structures on a per neuron/synapse basis. For resizing, we need to access information about things such as the length of indices, number of spikes, etc., how would we do this in abstract code? I think it would be a significant departure from our general abstract code syntax, I'm not sure this is really worth it.

**Comment by thesamovar** [comment]
Yeah I think you're right it wouldn't fit with abstract code syntax, it would have to be a separate code block. I think there is a potential payoff in terms of greatly reducing the complexity of the templates, but perhaps this is something we should put off to post 2.0.
